lms/djangoapps/experiments/flags.py

```diff
@@ -0,0 +1,133 @@
+"""
+Feature flag support for experiments
+"""
+
+import logging
+import pytz
+
+import dateutil
+from crum import get_current_request
+from edx_django_utils.cache import RequestCache
+
+from experiments.stable_bucketing import stable_bucketing_hash_group
+from openedx.core.djangoapps.waffle_utils import CourseWaffleFlag
+from track import segment
+
+log = logging.getLogger(__name__)
+
+
+class ExperimentWaffleFlag(CourseWaffleFlag):
+    """
+    ExperimentWaffleFlag handles logic around experimental bucketing and whitelisting.
+
+    You'll have one main flag that gates the experiment. This allows you to control the scope
+    of your experiment and always provides a quick kill switch.
+
+    But you'll also have smaller related flags that can force bucketing certain users into
+    specific buckets of your experiment. Those can be set using a waffle named like
+    "main_flag.BUCKET_NUM" (e.g. "course_experience.animated_exy.0") to force
+    users that pass the first main waffle check into a specific bucket experience.
+
+    If you pass this flag a course key, tracking calls to segment will be made per-course-run
+    (rather than one call overall) and will include the course key.
+
+    You can also control whether the experiment only affects future enrollments by setting
+    an ExperimentKeyValue model object with a key of 'enrollment_start' to the date of the
+    first enrollments that should be bucketed.
+
+    Bucket 0 is assumed to be the control bucket.
+
+    .. no_pii:
+    """
+    def __init__(self, waffle_namespace, flag_name, num_buckets=2, experiment_id=None, **kwargs):
+        super().__init__(waffle_namespace, flag_name, **kwargs)
+        self.num_buckets = num_buckets
+        self.experiment_id = experiment_id
+        self.bucket_flags = [
+            CourseWaffleFlag(waffle_namespace, '{}.{}'.format(flag_name, bucket), flag_undefined_default=False)
```

Comment:
I think a personal style, but just wanted to call out that CourseWaffleFlags have a default of being False so you don't _have_ to explicitly call that out, but being explicit never hurt.

Reply:
Fair. If I edit the code again, I'll make that change.

Reply:
Oh, I guess the wrapper code we have makes a distinction between the default (None) and False.

Reply:
Oh whaaaat? Alright, good to know. 

Don't love that the default is not a Boolean given that it is a boolean field, but fine.

Reply:
I think that it eventually reaches the same False conclusion, but does extra work to get there.